### PR TITLE
improved narrowing of generics

### LINKED ISSUE #674: `isinstance` checks on classes with generics should narrow to the generic's bound if it's covariant, and `Never` if it's contravariant
Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&reportImplicitOverride=true&code=MYGwhgzhAEBiD28DaAVAXNAlgOwC7QB9oJcAnAXTQFgAoae6AEwFMAzaVxACgmZFYA00eLgAWzUhngAjAFbNguAJTU6DdZnaYIOEmGzBmXEeNJCE8FbXU2GpZgDdmYEAH1cATwAORkxKXQAMRwiEgAqtgA1tjwAO7Y5EIQovAAriCM0NLMIcg4%2BEQkFEA)

```python
class Foo[T: int | str]:
    def foo(self, other: object):
        if isinstance(other, Foo):
            reveal_type(other) # Foo[Unknown], should be Foo[int | str]
```
if the generic is covariant it should narrow to `Foo[int | str]` (or whatever the generic's bound is), and if it's contravariant it should narrow to `Foo[Never]` i think

### PR DESCRIPTION
**Pull request by DetachHead**
- use a generic's bound when narrowing with `isinstance` if it's covariant, `Never` if it's contravariant
- creates a union of all possible combinations if it has constraints

fixes #674

### CODE REVIEW

File: packages/pyright-internal/src/tests/samples/typeNarrowingIsinstance6.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ class ChildB1(ParentB[_T2]):
 
 def func4(var: ParentB[int]):
     if isinstance(var, ChildB1):
-        reveal_type(var, expected_text="ChildB1[int]")
+        reveal_type(var, expected_text="<subclass of ParentB[int] and ChildB1[float]>")
```
Comment (beauxq) [inline packages/pyright-internal/src/tests/samples/typeNarrowingIsinstance6.py:54]: This doesn't seem like a good change. If we already have a narrower type var, it seems like it shouldn't widen it.
Do you know an example where narrowing to `ChildB1[int]` would be unsafe?

If it's just an issue of being difficult to implement, I think this change probably isn't bad enough to be a blocker.

**Comment by beauxq** [comment]
Thanks for working on this. I tried working on it for a while, but I got stuck on trying to understand the code.
I think this would be one of the most valuable features of basedpyright.

**Comment by DetachHead** [comment]
i ran into a difficult to solve edge case that led to me abandoning this change for now (see the failing test case) but i may give it another go in the future

File: packages/pyright-internal/src/analyzer/typeUtils.ts [hunk 2]
```diff
@@ -1053,9 +1053,35 @@ export function getTypeVarScopeIds(type: Type): TypeVarScopeId[] {
     return scopeIds;
 }
 
-// Specializes the class with "Unknown" type args (or the equivalent for ParamSpecs
-// or TypeVarTuples).
-export function specializeWithUnknownTypeArgs(type: ClassType, tupleClassType?: ClassType): ClassType {
+/**
+ * if the type we're narrowing is Any or Unknown, we don't want to specialize using the
+ * variance/bound for compatibility with less strictly typed code (cringe)
```
Comment (beauxq) [inline packages/pyright-internal/src/analyzer/typeUtils.ts:1058]: I would like `Any` and `Unknown` to specialize using the variance/bound for my code.

But I understand that for people who don't report `Unknown`, it would be more noisy.

@DetachHead  What do you think of a new configuration option:
strictGenericNarrowing

I think I could implement it.

Comment (beauxq) [reply packages/pyright-internal/src/analyzer/typeUtils.ts:1058]: Or can we just implement the new narrowing with `Any` / `Unknown` and see how many people complain about it?
It would be a bit of a waste to have an option when no one is asking for it.

Comment (DetachHead) [reply packages/pyright-internal/src/analyzer/typeUtils.ts:1058]: the option sounds good cuz i want to maintain backwards compatibility for pyright users where possible

**Comment by beauxq** [comment]
https://github.com/DetachHead/basedpyright/pull/885

You should pull from main here before looking at the diff there.

**Comment by DetachHead** [comment]
the primer identified this issue with constrained typevars:

```py
from typing import Literal


class Foo[T: (int, str)]:
    ...

def asdf(value: object):
    if isinstance(value, Foo):
        reveal_type(value) # `Foo[object]`, should be `Foo[int] | Foo[str]`
```

File: docs/configuration.md [hunk 3]
```diff
@@ -69,6 +73,10 @@ The following settings determine how different types should be evaluated.
 
 - <a name="disableBytesTypePromotions"></a> **disableBytesTypePromotions** [boolean]: Disables legacy behavior where `bytearray` and `memoryview` are considered subtypes of `bytes`. [PEP 688](https://peps.python.org/pep-0688/#no-special-meaning-for-bytes) deprecates this behavior, but this switch is provided to restore the older behavior.
 
+### basedpyright exclusive settings
+
+- <a name="improvedGenericNarrowing"></a> **improvedGenericNarrowing** [boolean]: When a type is narrowed in such a way that its type parameters are not known (eg. using an `isinstance` check), basedpyright will resolve the type parameter to the generic's bound or constraint instead of `Any`.
```
Comment (DetachHead) [inline docs/configuration.md:78]: add this option to the playground

File: docs/configuration.md [hunk 4]
```diff
@@ -69,6 +73,10 @@ The following settings determine how different types should be evaluated.
 
 - <a name="disableBytesTypePromotions"></a> **disableBytesTypePromotions** [boolean]: Disables legacy behavior where `bytearray` and `memoryview` are considered subtypes of `bytes`. [PEP 688](https://peps.python.org/pep-0688/#no-special-meaning-for-bytes) deprecates this behavior, but this switch is provided to restore the older behavior.
 
+### basedpyright exclusive settings
+
+- <a name="strictGenericNarrowing"></a> **strictGenericNarrowing** [boolean]: When a type is narrowed in such a way that its type parameters are not known (eg. using an `isinstance` check), basedpyright will resolve the type parameter to the generic's bound or constraint instead of `Any`.
```
Comment (beauxq) [inline docs/configuration.md:78]: I kind of liked the name "improved" when I saw you had that. but it's ok either way.

**Review by beauxq — APPROVED** [review]
I looked over all the code. Didn't see any problems.

File: docs/configuration.md (hunk 4 above)
Comment (DetachHead) [reply docs/configuration.md:78]: i decided to rename it to be consistent with other settings like `strictListInference`, `strictSetInference` etc

**Comment by jorenham** [comment]
this is fantastic
